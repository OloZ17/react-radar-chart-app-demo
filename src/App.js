import React, { useState } from 'react';
import { array } from 'prop-types';
import './App.css';
import { Skills, Requirements, Achievements, RadarChart } from './components';

const App = (props) => {
  const { _skills, _requirements, _achievements } = props;

  const initialSkillsState = [
    'Thing 1',
    'Thing 2',
    'Thing 3',
    'Thing 4',
    'Thing 5',
    'Thing 6',
  ];

  const initialRequirementsState = [2, 9, 3, 5, 2, 3];

  const initialAchievementsState = [2, 9, 3, 5, 2, 3];

  const [skills, setSkills] = _skills
    ? useState(_skills)
    : useState(initialSkillsState);

  const [requirements, setRequirements] = _requirements
    ? useState(_requirements)
    : useState(initialRequirementsState);

  const [achievements, setAchievements] = _achievements
    ? useState(_achievements)
    : useState(initialAchievementsState);

  const handleChangeSkills = (index, e) => {
    e.preventDefault();
    const newSkills = [...skills];
    newSkills[index] = e.target.value;
    setSkills(newSkills);
  };

  const removeSkill = (index) => {
    const newSkills = [...skills];
    newSkills.splice(index, 1);
    setSkills(newSkills);
    const newRequirements = [...requirements];
    newRequirements.splice(index, 1);
    setRequirements(newRequirements);
    const newAchievements = [...achievements];
    newAchievements.splice(index, 1);
    setAchievements(newAchievements);
  };

  const addSkill = () => {
    setSkills([...skills, 'DEFAULT']);
    setRequirements([...requirements, 0]);
    setAchievements([...achievements, 0]);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    alert(JSON.stringify(skills));
  };

  const handleChangeRequirement = (index, e) => {
    e.preventDefault();
    const newRequirements = [...requirements];
    newRequirements[index] = e.target.value;
    setRequirements(newRequirements);
  };

  const resetRequirement = (index) => {
    const newRequirements = [...requirements];
    newRequirements[index] = 0;
    setRequirements(newRequirements);
  };

  const handleChangeAchievement = (index, e) => {
    e.preventDefault();
    const newAchievementss = [...achievements];
    newAchievementss[index] = e.target.value;
    setAchievements(newAchievementss);
  };

  const resetAchievement = (index) => {
    const newAchievements = [...achievements];
    newAchievements[index] = 0;
    setAchievements(newAchievements);
  };

  const data = {
    labels: skills,
    datasets: [
      {
        label: 'Requirements',
        data: requirements,
        backgroundColor: 'rgba(255, 99, 132, 0.2)',
        borderColor: 'rgba(255, 99, 132, 1)',
        borderWidth: 1,
      },
      {
        label: 'Achievements',
        data: achievements,
        backgroundColor: 'rgba(12, 8, 255, 0.2)',
        borderColor: 'rgba(12, 8, 255, 1)',
        borderWidth: 1,
      },
    ],
  };

  return (
    <div className="App">
      <div className="container">
        <div className="row">
          <div className="col-md-6">
            <form onSubmit={handleSubmit}>
              <Skills
                skills={skills}
                handleChangeSkills={handleChangeSkills}
                addSkill={addSkill}
                removeSkill={removeSkill}
              />
              <div className="row">
                <Requirements
                  skills={skills}
                  requirements={requirements}
                  handleChangeRequirement={handleChangeRequirement}
                  resetRequirement={resetRequirement}
                />
                <Achievements
                  skills={skills}
                  achievements={achievements}
                  handleChangeAchievement={handleChangeAchievement}
                  resetAchievement={resetAchievement}
                />
              </div>
              <div className="button-section">
                <button className="btn btn-primary" type="submit">
                  Submit
                </button>
              </div>
            </form>
          </div>
          <RadarChart data={data} />
        </div>
      </div>
    </div>
  );
};

App.propTypes = {
  _skills: array,
  _requirements: array,
  _achievements: array,
};

export default App;
