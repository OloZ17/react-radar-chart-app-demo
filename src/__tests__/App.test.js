import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import App from '../App';
import userEvent from '@testing-library/user-event';

describe('Testing App() Component', () => {
  const { ResizeObserver } = window;

  beforeEach(() => {
    delete window.ResizeObserver;
    window.ResizeObserver = jest.fn().mockImplementation(() => ({
      observe: jest.fn(),
      unobserve: jest.fn(),
      disconnect: jest.fn(),
    }));
  });

  afterEach(() => {
    window.ResizeObserver = ResizeObserver;
    jest.restoreAllMocks();
  });

  it('should take a snapshot', () => {
    const { asFragment } = render(<App />);
    expect(asFragment(<App />)).toMatchSnapshot();
  });

  it('should add a skill', () => {
    render(<App />);
    const initialSize = screen.queryAllByRole('textbox').length;
    const addSkill_buttons = screen.queryAllByTestId('addSkill-button');
    fireEvent.click(addSkill_buttons[0]);
    const newSize = screen.queryAllByRole('textbox').length;
    expect(newSize) > initialSize;
  });

  it('should remove a skill', () => {
    render(<App />);
    const initialSize = screen.queryAllByRole('textbox').length;
    const removeSkill_buttons = screen.queryAllByTestId('removeSkill-button');
    fireEvent.click(removeSkill_buttons[0]);
    const newSize = screen.queryAllByRole('textbox').length;
    expect(newSize) < initialSize;
  });

  it('should handlechange skill', () => {
    render(<App _skills={['']} />);
    const inputs = screen.queryAllByRole('textbox');
    fireEvent.change(inputs[0], { target: { value: 'Test' } });
    expect(inputs[0]).toHaveValue('Test');
  });

  it('should reset requirements', () => {
    render(<App />);
    const button = screen.getAllByTestId('resetRequirement-button');
    fireEvent.click(button[0]);
    const inputs = screen.getAllByTestId('requirement-input');
    expect(inputs[0]).toHaveValue(0);
  });

  it('should handlechange requirements', () => {
    render(<App />);
    const inputs = screen.getAllByTestId('requirement-input');
    fireEvent.change(inputs[0], { target: { value: 100 } });
    expect(inputs[0]).toHaveValue(100);
  });

  it('should reset achievements', () => {
    render(<App />);
    const button = screen.getAllByTestId('resetAchievement-button');
    fireEvent.click(button[0]);
    const inputs = screen.getAllByTestId('achievement-input');
    expect(inputs[0]).toHaveValue(0);
  });

  it('should handlechange achievements', () => {
    render(<App />);
    const inputs = screen.getAllByTestId('achievement-input');
    fireEvent.change(inputs[0], { target: { value: 100 } });
    expect(inputs[0]).toHaveValue(100);
  });
});
