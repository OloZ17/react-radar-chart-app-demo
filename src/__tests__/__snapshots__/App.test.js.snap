// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing App() Component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <form>
            <div
              class="row"
            >
              <h3>
                Skills
              </h3>
              <div
                class="form-inline form-group mb-2"
              >
                <label>
                  Skill 1
                </label>
                <input
                  name="skill"
                  style="padding: 2px; text-align: left; margin-right: 5px; margin-left: 5px;"
                  type="text"
                  value="Thing 1"
                />
                <button
                  class="btn btn-danger"
                  data-testid="removeSkill-button"
                  type="button"
                >
                  Remove
                </button>
              </div>
              <div
                class="form-inline form-group mb-2"
              >
                <label>
                  Skill 2
                </label>
                <input
                  name="skill"
                  style="padding: 2px; text-align: left; margin-right: 5px; margin-left: 5px;"
                  type="text"
                  value="Thing 2"
                />
                <button
                  class="btn btn-danger"
                  data-testid="removeSkill-button"
                  type="button"
                >
                  Remove
                </button>
              </div>
              <div
                class="form-inline form-group mb-2"
              >
                <label>
                  Skill 3
                </label>
                <input
                  name="skill"
                  style="padding: 2px; text-align: left; margin-right: 5px; margin-left: 5px;"
                  type="text"
                  value="Thing 3"
                />
                <button
                  class="btn btn-danger"
                  data-testid="removeSkill-button"
                  type="button"
                >
                  Remove
                </button>
              </div>
              <div
                class="form-inline form-group mb-2"
              >
                <label>
                  Skill 4
                </label>
                <input
                  name="skill"
                  style="padding: 2px; text-align: left; margin-right: 5px; margin-left: 5px;"
                  type="text"
                  value="Thing 4"
                />
                <button
                  class="btn btn-danger"
                  data-testid="removeSkill-button"
                  type="button"
                >
                  Remove
                </button>
              </div>
              <div
                class="form-inline form-group mb-2"
              >
                <label>
                  Skill 5
                </label>
                <input
                  name="skill"
                  style="padding: 2px; text-align: left; margin-right: 5px; margin-left: 5px;"
                  type="text"
                  value="Thing 5"
                />
                <button
                  class="btn btn-danger"
                  data-testid="removeSkill-button"
                  type="button"
                >
                  Remove
                </button>
              </div>
              <div
                class="form-inline form-group mb-2"
              >
                <label>
                  Skill 6
                </label>
                <input
                  name="skill"
                  style="padding: 2px; text-align: left; margin-right: 5px; margin-left: 5px;"
                  type="text"
                  value="Thing 6"
                />
                <button
                  class="btn btn-danger"
                  data-testid="removeSkill-button"
                  type="button"
                >
                  Remove
                </button>
              </div>
              <div
                class="button-section"
              >
                <button
                  class="btn btn-primary"
                  data-testid="addSkill-button"
                  type="button"
                >
                  Add
                </button>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col-md-6"
              >
                <h3>
                  Requirements
                </h3>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 1
                  </label>
                  <input
                    data-testid="requirement-input"
                    name="requirement"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="2"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetRequirement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 2
                  </label>
                  <input
                    data-testid="requirement-input"
                    name="requirement"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="9"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetRequirement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 3
                  </label>
                  <input
                    data-testid="requirement-input"
                    name="requirement"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="3"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetRequirement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 4
                  </label>
                  <input
                    data-testid="requirement-input"
                    name="requirement"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="5"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetRequirement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 5
                  </label>
                  <input
                    data-testid="requirement-input"
                    name="requirement"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="2"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetRequirement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 6
                  </label>
                  <input
                    data-testid="requirement-input"
                    name="requirement"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="3"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetRequirement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
              </div>
              <div
                class="col-md-6"
              >
                <h3>
                  Achievements
                </h3>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 1
                  </label>
                  <input
                    data-testid="achievement-input"
                    name="achievements"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="2"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetAchievement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 2
                  </label>
                  <input
                    data-testid="achievement-input"
                    name="achievements"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="9"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetAchievement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 3
                  </label>
                  <input
                    data-testid="achievement-input"
                    name="achievements"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="3"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetAchievement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 4
                  </label>
                  <input
                    data-testid="achievement-input"
                    name="achievements"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="5"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetAchievement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 5
                  </label>
                  <input
                    data-testid="achievement-input"
                    name="achievements"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="2"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetAchievement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
                <div
                  class="form-inline form-group mb-2"
                >
                  <label>
                    Thing 6
                  </label>
                  <input
                    data-testid="achievement-input"
                    name="achievements"
                    style="padding: 2px; width: 50px; text-align: center; margin-right: 5px; margin-left: 5px;"
                    type="number"
                    value="3"
                  />
                  <button
                    class="btn btn-warning"
                    data-testid="resetAchievement-button"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
              </div>
            </div>
            <div
              class="button-section"
            >
              <button
                class="btn btn-primary"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
        <div
          class="col-md-6"
        >
          <canvas
            height="0"
            role="img"
            style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
            width="0"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
